browser/ui/views/playlist/playlist_action_icon_view.cc

```diff
@@ -0,0 +1,117 @@
+#include "ui/base/metadata/metadata_impl_macros.h"
+
+PlaylistActionIconView::PlaylistActionIconView(
+    CommandUpdater* command_updater,
+    Browser* browser,
+    IconLabelBubbleView::Delegate* icon_label_bubble_delegate,
+    PageActionIconView::Delegate* page_action_icon_delegate)
+    : PageActionIconView(command_updater,
+                         0,
+                         icon_label_bubble_delegate,
+                         page_action_icon_delegate,
+                         "PlaylistActionIconView",
+                         /*ephemeral=*/false) {
+  SetVisible(false);
+}
+
+PlaylistActionIconView::~PlaylistActionIconView() = default;
+
+views::BubbleDialogDelegate* PlaylistActionIconView::GetBubble() const {
+  return PlaylistActionBubbleView::GetBubble();
+}
+
+void PlaylistActionIconView::OnExecuting(ExecuteSource execute_source) {
+  if (PlaylistActionBubbleView::IsShowingBubble()) {
+    return;
+  }
+
+  DCHECK_EQ(current_web_contents_, GetWebContents());
+  PlaylistActionBubbleView::ShowBubble(this,
+                                       base::to_address(current_web_contents_));
+}
+
+const gfx::VectorIcon& PlaylistActionIconView::GetVectorIcon() const {
+  return state_ == kAdded ? kPlaylistActionAddedIcon : kPlaylistActionIcon;
+}
+
+void PlaylistActionIconView::UpdateImpl() {
+  if (auto old_contents =
+          std::exchange(current_web_contents_, GetWebContents());
+      old_contents == current_web_contents_) {
+    return;
+  }
+
+  if (PlaylistActionBubbleView::IsShowingBubble()) {
+    PlaylistActionBubbleView::CloseBubble();
+  }
+
+  auto* playlist_tab_helper = this->playlist_tab_helper();
+  if (!playlist_tab_helper) {
+    UpdateState(0, 0);
+    return;
+  }
+
+  saved_items_changed_subscription_ =
+      playlist_tab_helper->RegisterSavedItemsChangedCallback(
+          base::BindRepeating(&PlaylistActionIconView::OnSavedItemChanged,
+                              base::Unretained(this)));
+  found_items_changed_subscription_ =
+      playlist_tab_helper->RegisterFoundItemsChangedCallback(
+          base::BindRepeating(&PlaylistActionIconView::OnFoundItemChanged,
+                              base::Unretained(this)));
+
+  UpdateState(playlist_tab_helper->saved_items().size(),
+              playlist_tab_helper->found_items().size());
+}
+
+void PlaylistActionIconView::OnSavedItemChanged(
+    const std::vector<playlist::mojom::PlaylistItemPtr>& saved_items) {
+  auto* playlist_tab_helper = this->playlist_tab_helper();
+  CHECK(playlist_tab_helper);
+
+  UpdateState(saved_items.size(), playlist_tab_helper->found_items().size());
+}
+
+void PlaylistActionIconView::OnFoundItemChanged(
+    const std::vector<playlist::mojom::PlaylistItemPtr>& found_items) {
+  auto* playlist_tab_helper = this->playlist_tab_helper();
+  CHECK(playlist_tab_helper);
+
+  UpdateState(playlist_tab_helper->saved_items().size(), found_items.size());
+}
+
+void PlaylistActionIconView::UpdateState(int saved_items_count,
```

Comment:
Mega optional nit: 
I wonder if we should just pass in booleans for `has_saved` and `found_items` instead of the count?